lsp/nls/src/requests/hover.rs

```diff
@@ -156,57 +160,47 @@ pub fn handle(
-        // use this annotation insted.
-        let mut type_annots: Vec<_> = hover
+        // Collect all the type and contract annotations we can find. (We don't distinguish between them.)
+        let mut annotations: Vec<_> = hover
             .metadata
             .iter()
-            .filter_map(|m| Some(m.annotation.typ.as_ref()?.typ.to_string()))
+            .flat_map(|m| {
+                let maybe_typ = m.annotation.typ.as_ref().map(|typ| typ.typ.to_string());
+                let contracts = m
+                    .annotation
+                    .contracts
+                    .iter()
+                    .map(|c| c.label.typ.to_string());
+                maybe_typ.into_iter().chain(contracts)
+            })
+            .chain(
+                hover
+                    .values
+                    .iter()
+                    .flat_map(annotated_contracts)
+                    .map(|contract| contract.label.typ.to_string()),
+            )
             .collect();
+        annotations.sort();
+        annotations.dedup();
 
         let ty = hover
             .ty
             .as_ref()
             .map(Type::to_string)
-            // Unclear whether it's useful to report `Dyn` all the time when there's no type found,
-            // but it matches the old behavior.
             .unwrap_or_else(|| "Dyn".to_owned());
 
-        // If the type is `Dyn`, and we can find a type annotation somewhere in the metadata, we
-        // use the latter instead, which will be more precise.
-        let ty = if ty == "Dyn" {
-            // Ordering isn't meaningful here: metadata are aggregated from merged values (and
-            // merge is commutative). This list will also be sorted for deduplication later anyway.
-            // So we just pop the last one.
-            type_annots.pop().unwrap_or(ty)
-        } else {
-            // If the type is both statically known and present as an annotation, we don't want to
-            // report a second time with the other contracts, so we remove it from the list.
-            //
-            // Note that there might be duplicates, and we need to remove them all, hence the
-            // `retain` (instead of a potentially more performant `iter().position(..)` followed by
-            // `swap_remove`).
-            type_annots.retain(|annot| annot != &ty);
-
-            ty
-        };
-
-        contents.push(nickel_string(ty));
-
-        let mut contracts: Vec<_> = hover
-            .metadata
-            .iter()
-            .flat_map(|m| &m.annotation.contracts)
-            .chain(hover.values.iter().flat_map(annotated_contracts))
-            .map(|contract| contract.label.typ.to_string())
-            .chain(type_annots)
-            .collect();
+        // There's no point in repeating the static type in the annotations.
+        if let Some(idx) = annotations.iter().position(|a| a == &ty) {
+            annotations.remove(idx);
+        }
```

Comment:
I don't understand why we remove the static type from the annotations.  `annotations` is just the list of the type, followed by all contracts, sorted and deduped, right? So, I would expect to find the static type in it, and that it's intentional. Why remove it?

Reply:
Ok, I get it. You include the static type in the annotations so that it's deduped (if there is the same as a contract), but you still keep the distinction static type/contracts, so you display the static type separately, and don't want to repeat it in the annotations.

Which I think is the right behavior (the static type is special in that it's the one selected e.g. by the typechecker). Maybe you could add a short mention explaining this (we don't distinguish between types and contracts for the dedup but we will in the LSP answer) to the comment attached to `annotations` ?

Reply:
Yes, exactly. I've added a comment so hopefully it's less mysterious now